[FEATURE] local affine (un-)banded alignment

joergi-w opened the pull request:
This PR adds the implementation and a test for the Smith Waterman algorithm for local alignments.

TODO: performance test

Solves issue #321.

rrahn commented on include/seqan3/alignment/pairwise/alignment_algorithm.hpp:
```diff
@@ -525,16 +525,19 @@ class alignment_algorithm :
-                std::ranges::advance(it, (gap_elem.position - normalise) + offset);
-                aligned_sequence.insert(it, gap_elem.size, gap{});
-                offset += gap_elem.size;
+                assert(normalise <= gap_segments[0].position);
```
And revert the additional if condition. It should work then for empty ranges anyway

rrahn commented on include/seqan3/alignment/configuration/align_config_mode.hpp:
```diff
@@ -48,14 +58,26 @@ inline constexpr detail::global_alignment_type global_alignment;
+ *
+ * \snippet snippet/alignment/configuration/align_cfg_mode_example.cpp local
+ */
+inline constexpr detail::local_alignment_type local_alignment;
```
please put this into the seqan3 namespace above and change on all places.

joergi-w replied:
Okay!

joergi-w commented:
Rebased on upstream/master.

rrahn commented on test/performance/alignment/local_affine_alignment_benchmark.cpp:
```diff
@@ -0,0 +1,277 @@
+using namespace seqan3;
+
+template <typename alphabet_t>
+auto generate_sequence_seqan3(size_t const len = 500,
```
This should be refactored and DRYied out. There is an open PR #755 which does extracts this part already. We are currently working on getting this in and then it would be good to use the global sequence generation for the local and the global alignment.

rrahn commented:
@joergi-w can you please rebase on master and then remove the sequence generator and use the global one  introduced by #755. I'll give then the final review.

joergi-w commented:
Rebased on master and updated the performance tests. Waiting for #903.

Open TODOs:
* The traceback path is different from SeqAn2: SeqAn3 uses diagonal edges first, SeqAn2 prefers vertical edges (see dna4_03 in `local_affine_unbanded_test`).
* Write tests for banded local alignment.
* Remove commit 1465975 after PR #903 is merged and rebase again.

joergi-w commented:
We have tests for the banded local alignment now and I verified all tests with SeqAn2 local alignments. We show a better error message, whenever the band ranges exclude the whole matrix computation. Band bounds that exceed the sequence sizes had led to huge dp matrix allocation and invalid indices, which was fixed in the last commit. 

This PR is ready for another review; merging has to wait for #903.

rrahn commented on include/seqan3/alignment/pairwise/policy/banded_score_dp_matrix_policy.hpp:
```diff
@@ -91,6 +91,10 @@ class banded_score_dp_matrix_policy :
                                            static_cast<int_fast32_t>(0)));
 
+        // If the band is wider than the sequence length, limit the band width.
+        band_column_index = std::min(band_column_index, static_cast<uint_fast32_t>(dimension_second_range - 1));
```
This should actually not impose a problem. Why this change?

joergi-w replied:
It does impose a problem, because if the user specifies a band width larger than sequence length, the `band_column_index` is out of bounds. 
```console
Assertion failed: (current_column_index > band_column_index), function second_range_begin_offset, file /Users/jwinkler/repo/seqan3/include/seqan3/alignment/pairwise/policy/banded_score_dp_matrix_policy.hpp, line 174.
```
Furthermore, it will allocate a too large DP matrix a few lines below.

rrahn replied:
Strange, I need to investigate this, because basically it should not matter.

joergi-w commented on include/seqan3/alignment/pairwise/alignment_configurator.hpp:
```diff
@@ -504,38 +507,46 @@ constexpr function_wrapper_t alignment_configurator::configure_free_ends_initial
+            // If possible use static information.
+            if constexpr (align_ends_cfg_t::template is_static<2>())
+            {
+                return configure_leading_both(first,
```
Yes, good idea!

joergi-w commented:
Rebased on master to include PR #903.

joergi-w commented on include/seqan3/alignment/configuration/detail.hpp:
```diff
@@ -45,14 +46,15 @@ template <>
+        { 1, 1, 0, 1, 1, 1, 1, 1}, // 2: gap
+        { 1, 1, 1, 0, 0, 1, 1, 1}, // 3: global
+        { 0, 1, 1, 0, 0, 0, 1, 1}, // 4: local
+        { 1, 1, 1, 1, 0, 0, 1, 1}, // 5: max_error
```
This config limits the number of allowed errors across all error types (mismatches, insertions, deletions) in global alignment. The local alignment anyway finds the areas of high similarity, so it is not useful to constrain the number of errors here.

smehringer replied:
but maybe I want to have only local alignments with up to `x` errors? :grin:

rrahn replied:
In general max_errors does only work with the edit distance computation. Edit distance is only available for global alignment anyway. In the normal alignment we can not easily keep track of number of errors and we don't need it for now. Unfortunately, the combinations can not be displayed but it is documentated and there are runtime checks.

joergi-w commented on include/seqan3/alignment/pairwise/alignment_configurator.hpp:
```diff
@@ -504,38 +507,46 @@ constexpr function_wrapper_t alignment_configurator::configure_free_ends_initial
+    {
+        // This lambda determines the initialisation configuration for the second sequence given
+        // the leading gap property for it.
+        auto configure_leading_second = [&](auto first) constexpr
```
Then I'm going to change the space consistently in the whole file...

joergi-w commented on include/seqan3/alignment/configuration/detail.hpp:
```diff
@@ -26,6 +26,7 @@ enum struct align_config_id : uint8_t
     band, //!< ID for the \ref seqan3::align_cfg::band "band" option.
     gap, //!< ID for the \ref seqan3::align_cfg::gap "gap" option.
     global, //!< ID for the \ref seqan3::global_alignment "global alignment" option.
+    local, //!< ID for the \ref seqan3::local_alignment "local alignment" option.
```
Sure!

joergi-w commented on include/seqan3/alignment/pairwise/alignment_configurator.hpp:
```diff
@@ -504,38 +507,46 @@ constexpr function_wrapper_t alignment_configurator::configure_free_ends_initial
+        auto configure_leading_second = [&](auto first) constexpr
+        {
+            // If possible use static information.
+            if constexpr (align_ends_cfg_t::template is_static<2>())
```
Indices 0 and 2 are used for free leading gaps; 1 and 3 are used for free trailing gaps (in the function below).

rrahn replied:
This is semantically correct.

rrahn commented on include/seqan3/alignment/pairwise/policy/affine_gap_banded_policy.hpp:
```diff
@@ -90,13 +94,17 @@ class affine_gap_banded_policy : public affine_gap_policy<affine_gap_banded_poli
             main_score = (main_score < vt_score) ? vt_score : main_score;
             main_score = (main_score < prev_hz_score) ? prev_hz_score : main_score;
+            if constexpr (align_local_t::value)
+                main_score = (main_score < 0) ? 0 : main_score;
```
It can be modeled by simd types as well.

joergi-w replied:
According to @rrahn it makes a difference, so I trust him ;)

joergi-w commented:
I rebased on master and cleaned the git history.
